### PR TITLE
Feature Object detection

### PR DESCRIPTION
**Pull request by Animatory**

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -24,5 +24,8 @@ dependencies:
       - hydra-core==1.2.0
       - opencv-python==4.6.0.*
       - ranx==0.2.8
-      - timm==0.6.5
-      - pycocotools==2.0.4
```
Comment (PososikTeam) [inline environment.yml:28]: Is mmdet has pycocotools?

File: torchok/models/heads/detection/yolov3.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                scores,
+                cfg['score_thr'],
+                dict(cfg['nms']),
+                cfg['max_per_img'],
+                score_factors=objectness)
+            det_results.append(dict(bboxes=det_bboxes, labels=det_labels))
+        return det_results
+
+    @force_fp32(apply_to=('pred_maps',))
+    def loss(self, joint_loss: JointLoss, pred_maps: List[Tensor], gt_bboxes: List[Tensor], gt_labels: List[Tensor],
```
Comment (PososikTeam) [inline torchok/models/heads/detection/yolov3.py:147]: In MMDetection loss and get_bboxes method has already writen. Why we need to overwrite this method here?

File: environment.yml (hunk 1 above)
Comment (Animatory) [reply environment.yml:28]: Yes

File: torchok/models/heads/detection/yolov3.py (hunk 2 above)
Comment (Animatory) [reply torchok/models/heads/detection/yolov3.py:147]: Because we have JointLoss that cannot be adapted otherwise from mmdet

File: torchok/metrics/detection.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import numpy as np
+from mmdet.core.evaluation.mean_ap import eval_map
+from torch import Tensor
+from torchmetrics import Metric
+
+from torchok.constructor import METRICS
+
+
+@METRICS.register_class
+class MMDetectionMAP(Metric):
```
Comment (PososikTeam) [inline torchok/metrics/detection.py:12]: Has it problems with memory?

File: torchok/models/backbones/__init__.py [hunk 4]
```diff
@@ -6,3 +8,7 @@
 import torchok.models.backbones.mobilenetv3
 import torchok.models.backbones.resnet
 import torchok.models.backbones.swin
+
+has_mmcv = importlib.util.find_spec("mmcv")
```
Comment (PososikTeam) [inline torchok/models/backbones/__init__.py:12]: Why we need this? I think just import is ok

File: torchok/models/heads/detection/__init__.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import importlib
+
+has_mmcv = importlib.util.find_spec("mmcv")
```
Comment (PososikTeam) [inline torchok/models/heads/detection/__init__.py:3]: I think just import is ok

File: torchok/tasks/detection.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        output = self.forward_with_gt(batch)
+        output['prediction'] = self.bbox_head.get_bboxes(**output)
+        output['target'] = [dict(bboxes=bb, labels=la) for bb, la in zip(output['gt_bboxes'], output['gt_labels'])]
+        self.metrics_manager.update(Phase.TEST, **output)
+
+    def predict_step(self, batch: Dict[str, Union[torch.Tensor, int]], batch_idx: int) -> Dict[str, torch.Tensor]:
+        """Complete predict loop."""
+        output = self.forward_with_gt(batch)
+        output['prediction'] = self.bbox_head.get_bboxes(**output)
+        output['target'] = [dict(bboxes=bb, labels=la) for bb, la in zip(output['gt_bboxes'], output['gt_labels'])]
```
Comment (PososikTeam) [inline torchok/tasks/detection.py:105]: Need to check if output has gt_bboxes and gt_labels or remove this line, because it's predict

File: torchok/tasks/detection.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__(hparams)
+
+        # BACKBONE
+        backbone_name = self._hparams.task.params.get('backbone_name')
+        backbones_params = self._hparams.task.params.get('backbone_params', dict())
+        self.backbone = BACKBONES.get(backbone_name)(**backbones_params)
+        self.num_scales = self._hparams.task.params.get('num_scales', len(self.backbone.out_encoder_channels))
```
Comment (PososikTeam) [inline torchok/tasks/detection.py:27]: Need comment that is num_scales, and why it can be less than backbone features

Comment (PososikTeam) [reply torchok/tasks/detection.py:27]: It's also bad that we can't write parameter like `self.num_scales` in documentation, need to discuss on torchok weekly ***

File: torchok/tasks/detection.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        x = self.backbone.forward_features(x)[-self.num_scales:]
+        with torch.cuda.amp.autocast(dtype=torch.float32):
+            x = self.neck(x)
+        x = self.bbox_head(x)
+        return x
+
+    def forward_with_gt(self, batch: Dict[str, Union[torch.Tensor, int]]) -> Dict[str, Any]:
+        """Forward with ground truth labels."""
+        input_data = batch.get('image')
+        features = self.backbone.forward_features(input_data)[-self.num_scales:]
```
Comment (PososikTeam) [inline torchok/tasks/detection.py:52]: Our forward feature reurn [x] + features, is mmdet also need x?

File: torchok/models/heads/detection/fcos.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+        for k, v in kwargs.items():
+            if isinstance(v, DictConfig):
+                kwargs[k] = ConfigDict(OmegaConf.to_container(v, resolve=True))
+        AnchorFreeHead.__init__(self, num_classes, in_channels, **kwargs)
+
+        self.init_weights()
+
+    @staticmethod
+    def format_dict(head_output):
```
Comment (PososikTeam) [inline torchok/models/heads/detection/fcos.py:105]: may be replace this method to task?

File: torchok/models/heads/detection/fcos.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        num_pos = max(reduce_mean(num_pos), 1.0)
+
+        pos_bbox_preds = flatten_bbox_preds[pos_inds]
+        pos_centerness = flatten_centerness[pos_inds]
+        pos_bbox_targets = flatten_bbox_targets[pos_inds]
+        pos_centerness_targets = self.centerness_target(pos_bbox_targets)
+        # centerness weighted iou loss
+        centerness_denorm = max(reduce_mean(pos_centerness_targets.sum().detach()), 1e-6)
+
+        # if len(pos_inds) > 0:
```
Comment (PososikTeam) [inline torchok/models/heads/detection/fcos.py:173]: MMDet code in this section is
```
        if len(pos_inds) > 0:
            pos_points = flatten_points[pos_inds]
            pos_decoded_bbox_preds = self.bbox_coder.decode(
                pos_points, pos_bbox_preds)
            pos_decoded_target_preds = self.bbox_coder.decode(
                pos_points, pos_bbox_targets)
            loss_bbox = self.loss_bbox(
                pos_decoded_bbox_preds,
                pos_decoded_target_preds,
                weight=pos_centerness_targets,
                avg_factor=centerness_denorm)
            loss_centerness = self.loss_centerness(
                pos_centerness, pos_centerness_targets, avg_factor=num_pos)
        else:
            loss_bbox = pos_bbox_preds.sum()
            loss_centerness = pos_centerness.sum()
```
So the question is can losses calculate loss for empty tensors as inputs?

Comment (PososikTeam) [reply torchok/models/heads/detection/fcos.py:173]: If so remove comment

File: torchok/models/heads/detection/fcos.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+        Returns:
+            list[list[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is an (n, 5) tensor, where the first 4 columns
+                are bounding box positions (tl_x, tl_y, br_x, br_y) and the
+                5-th column is a score between 0 and 1. The second item is a
+                (n,) tensor where each item is the predicted class label of
+                the corresponding box.
+        """
+        pseudo_meta = [dict(img_shape=image_shape, scale_factor=1) for i in range(len(cls_scores[0]))]
```
Comment (PososikTeam) [inline torchok/models/heads/detection/fcos.py:214]: May you explain me why `get_bboxes` work in mmdet fcos? because default parameter for `img_metas` is None and they have a code line 
```for img_id in range(len(img_metas)):``` so it must fall.

File: examples/notebooks/CheckTransforms.ipynb [hunk 12]
```diff
@@ -3,11 +3,7 @@
   {
```
Comment (PososikTeam) [inline examples/notebooks/CheckTransforms.ipynb:3]: Please remember me, how you solve issue with bboxes out of image shape

**Review by PososikTeam — CHANGES_REQUESTED** [review]
Please, answer my questions and think about suggestions. Also i think you can change readme, because a lot of things in roadmap has already done

File: examples/notebooks/CheckTransforms.ipynb (hunk 12 above)
Comment (Animatory) [reply examples/notebooks/CheckTransforms.ipynb:3]: Check function `DetectionDataset.filter_bboxes' in torchok/data/datasets/detection/detection.py

File: torchok/metrics/detection.py (hunk 3 above)
Comment (Animatory) [reply torchok/metrics/detection.py:12]: The problem with memory is in another metric `CocoEvalMAP`. But this problem may be reduced by using `compute_on_step: false` parameter

File: torchok/models/backbones/__init__.py (hunk 4 above)
Comment (Animatory) [reply torchok/models/backbones/__init__.py:12]: It is for cases when you do not install mmcv. Without it, CI/CD will fail

File: torchok/tasks/detection.py (hunk 6 above)
Comment (Animatory) [reply torchok/tasks/detection.py:105]: Fixed

File: torchok/tasks/detection.py (hunk 7 above)
Comment (Animatory) [reply torchok/tasks/detection.py:27]: Agree

File: torchok/tasks/detection.py (hunk 8 above)
Comment (Animatory) [reply torchok/tasks/detection.py:52]: No, here we take the features from the latest backbone stages like `[layer2, layer3, layer4]`

File: torchok/models/heads/detection/fcos.py (hunk 9 above)
Comment (Animatory) [reply torchok/models/heads/detection/fcos.py:105]: No, this method is special for each detection head.

File: torchok/models/heads/detection/fcos.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+                 num_classes,
+                 in_channels,
+                 regress_ranges=((-1, 64), (64, 128), (128, 256), (256, 512), (512, INF)),
+                 center_sampling=False,
+                 center_sample_radius=1.5,
+                 norm_on_bbox=False,
+                 centerness_on_reg=False,
+                 norm_cfg=dict(type='GN', num_groups=32, requires_grad=True),
+                 conv_cfg=None,
+                 train_cfg=None,
```
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/models/heads/detection/fcos.py:73]: These fields are taken from MMDet and are set up inside the MMDet code. I will think about how to add documentation for that

File: torchok/models/heads/detection/fcos.py (hunk 11 above)
Comment (Animatory) [reply torchok/models/heads/detection/fcos.py:214]: In mmdet this function cannot work without `img_metas`. It is a trick that the default value is None. But inside their code the meta is used only to get image shape. Therefore in my implementation I initialize it with base image shape.

File: torchok/models/heads/detection/fcos.py (hunk 10 above)
Comment (Animatory) [reply torchok/models/heads/detection/fcos.py:173]: It works with `len(pos_inds) == 0`

**Review by PososikTeam — APPROVED** [review]
